File: tests/integration/api-gateway.test.ts
/**
 * System Integration Tests – API Gateway Routing
 *
 * Verifies that the API Gateway (Nginx) correctly proxies
 * SOAP requests to the soap-processor and returns proper responses.
 * Tests the integration-layer pipeline from the gateway entry point
 * through to the SQS queue.
 */

import { describe, it, expect } from "vitest";
import { API_GATEWAY_URL } from "../helpers";

const VALID_SOAP = `<?xml version="1.0" encoding="UTF-8"?>
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
                  xmlns:req="http://example.com/integration/request">
  <soapenv:Header/>
  <soapenv:Body>
    <req:ProcessRequest>
      <req:RequestId>TEST-GATEWAY-001</req:RequestId>
      <req:Action>TestAction</req:Action>
      <req:Description>API Gateway routing test</req:Description>
    </req:ProcessRequest>
  </soapenv:Body>
</soapenv:Envelope>`;

describe("API Gateway – SOAP Routing", () => {
  it("proxies a valid SOAP request and returns HTTP 202", async () => {
    const res = await fetch(`${API_GATEWAY_URL}/soap`, {
      method: "POST",
      headers: { "Content-Type": "text/xml; charset=utf-8" },
      body: VALID_SOAP,
    });

    expect(res.status).toBe(202);
    const body = await res.text();
    expect(body).toContain("<res:Status>Accepted</res:Status>");
    expect(body).toContain("TEST-GATEWAY-001");
  });

  it("returns a SOAP error for invalid XML", async () => {
    const res = await fetch(`${API_GATEWAY_URL}/soap`, {
      method: "POST",
      headers: { "Content-Type": "text/xml; charset=utf-8" },
      body: "this is not xml",
    });

    // soap-processor returns 400 for missing SOAP Body
    expect(res.status).toBe(400);
    const body = await res.text();
    expect(body).toContain("Missing SOAP Body");
  });

  it("returns a SOAP error for XML without SOAP envelope", async () => {
    const res = await fetch(`${API_GATEWAY_URL}/soap`, {
      method: "POST",
      headers: { "Content-Type": "text/xml; charset=utf-8" },
      body: `<?xml version="1.0"?><root><data>test</data></root>`,
    });

    expect(res.status).toBe(400);
    const body = await res.text();
    expect(body).toContain("Missing SOAP Body");
  });

  it("generates unique RequestIds in responses for consecutive requests", async () => {
    const ids: string[] = [];

    for (let i = 0; i < 3; i++) {
      const soap = VALID_SOAP.replace(
        "TEST-GATEWAY-001",
        `TEST-UNIQUE-${Date.now()}-${i}`
      );
      const res = await fetch(`${API_GATEWAY_URL}/soap`, {
        method: "POST",
        headers: { "Content-Type": "text/xml; charset=utf-8" },
        body: soap,
      });
      const body = await res.text();
      const match = body.match(/<res:RequestId>(.+?)<\/res:RequestId>/);
      if (match) ids.push(match[1]);
    }

    // All three RequestIds should be different
    expect(new Set(ids).size).toBe(3);
  });
});

File: playwright-java-framework/src/test/java/com/integration/tests/integration/ApiGatewayTest.java
package com.integration.tests.integration;

import com.integration.framework.api.ApiClient;
import com.integration.framework.api.SoapClient;
import com.integration.framework.api.SoapClient.SoapResponse;
import com.integration.tests.base.BaseTest;
import org.junit.jupiter.api.*;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * System Integration Tests – API Gateway Routing
 *
 * Verifies that the API Gateway (Nginx) correctly proxies
 * SOAP requests to the soap-processor and returns proper responses.
 */
@Tag("integration")
@DisplayName("API Gateway – SOAP Routing")
class ApiGatewayTest extends BaseTest {

    private static final String VALID_SOAP = """
            <?xml version="1.0" encoding="UTF-8"?>
            <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
                              xmlns:req="http://example.com/integration/request">
              <soapenv:Header/>
              <soapenv:Body>
                <req:ProcessRequest>
                  <req:RequestId>TEST-GATEWAY-001</req:RequestId>
                  <req:Action>TestAction</req:Action>
                  <req:Description>API Gateway routing test</req:Description>
                </req:ProcessRequest>
              </soapenv:Body>
            </soapenv:Envelope>""";

    @Test
    @DisplayName("proxies a valid SOAP request and returns HTTP 202")
    void validSoapReturns202() {
        try (SoapClient soap = new SoapClient(apiGatewayUrl())) {
            SoapResponse res = soap.send(VALID_SOAP);

            assertThat(res.status()).isEqualTo(202);
            assertThat(res.body()).contains("<res:Status>Accepted</res:Status>");
            assertThat(res.body()).contains("TEST-GATEWAY-001");
        }
    }

    @Test
    @DisplayName("returns a SOAP error for invalid XML")
    void invalidXmlReturns400() {
        try (SoapClient soap = new SoapClient(apiGatewayUrl())) {
            SoapResponse res = soap.send("this is not xml");

            assertThat(res.status()).isEqualTo(400);
            assertThat(res.body()).contains("Missing SOAP Body");
        }
    }

    @Test
    @DisplayName("returns a SOAP error for XML without SOAP envelope")
    void missingEnvelopeReturns400() {
        try (SoapClient soap = new SoapClient(apiGatewayUrl())) {
            SoapResponse res = soap.send(
                    "<?xml version=\"1.0\"?><root><data>test</data></root>");

            assertThat(res.status()).isEqualTo(400);
            assertThat(res.body()).contains("Missing SOAP Body");
        }
    }

    @Test
    @DisplayName("generates unique RequestIds in responses for consecutive requests")
    void uniqueRequestIds() {
        Pattern idPattern = Pattern.compile("<res:RequestId>(.+?)</res:RequestId>");
        Set<String> ids = new HashSet<>();

        try (SoapClient soap = new SoapClient(apiGatewayUrl())) {
            for (int i = 0; i < 3; i++) {
                String body = VALID_SOAP.replace(
                        "TEST-GATEWAY-001",
                        "TEST-UNIQUE-" + System.currentTimeMillis() + "-" + i);
                SoapResponse res = soap.send(body);
                Matcher m = idPattern.matcher(res.body());
                if (m.find()) ids.add(m.group(1));
            }
        }

        assertThat(ids).hasSize(3);
    }
}
